src/btree/bt_delete.c

```diff
@@ -293,19 +293,43 @@ __wt_delete_page_rollback(WT_SESSION_IMPL *session, WT_REF *ref)
  *     IDs in the delete info. Since we're called at the end of recovery there's no need to lock the
  *     ref or worry about races.
  */
-static void
+static int
 __delete_redo_window_cleanup_internal(WT_SESSION_IMPL *session, WT_REF *ref)
 {
     WT_REF *child;
+    uint64_t sleep_usecs, yield_count;
+    bool busy;
 
     WT_ASSERT(session, F_ISSET(ref, WT_REF_FLAG_INTERNAL));
     if (ref->page != NULL) {
         WT_INTL_FOREACH_BEGIN (session, ref->page, child) {
-            if (child->state == WT_REF_DELETED && child->page_del != NULL)
+            /*
+             * We used to do the cleanup when the state of the child was WT_REF_DELETED, but it can
+             * also be loaded in memory if the page has been read before, thus we always clean the
+             * page_del structure if it has data.
+             */
+            if (child->page_del != NULL) {
                 __cell_redo_page_del_cleanup(session, ref->page->dsk, child->page_del);
+                /* Move back the page to WT_REF_DELETED by marking it for eviction. */
+                if (child->state == WT_REF_MEM) {
+                    for (sleep_usecs = yield_count = 0;;) {
+                        WT_RET(__wt_hazard_set(session, child, &busy));
+                        if (!busy) {
+                            WT_RET_BUSY_OK(__wt_page_release_evict(session, child, 0));
+                            break;
+                        }
+                        /* Wait some time to hopefully have access to the page. */
+                        __wt_spin_backoff(&yield_count, &sleep_usecs);
+                        WT_STAT_CONN_INCRV(session, fast_truncate_page_blocked, sleep_usecs);
+                        if (++yield_count > WT_THOUSAND * 10)
+                            return (EBUSY);
```

Comment:
If we return ebusy, what will happen afterwards?

Reply:
Will that increase the recovery time?

Reply:
> Returning EBUSY will lead to recovery failure. Restarting the server again can pass the recovery as we have already updated the base write generation numbers, so all the transaction ids are automatically reset during the page read.

In the worst-case scenario, since we are doing this for N pages, this means we could return EBUSY N times and need to restart the server N times, is that correct? If yes, I am afraid we need to find a different approach.

Reply:
The statlog-server in parallel to recovery can hold a maximum of 2 pages, it is very difficult that it holds the same page that we are trying to evict after the second restart.